README.md

# OVERVIEW 

Signed kernel modules provide a mechanism for the kernel to verify the integrity of a module.
This provides the tools needed to build a kernel with support for signed modules.

## Contents

    1. Introduction
    2. How to sign kernel modules using a custom kernel
    3. Summary of what needs to be done
    4. Kernel configuration
        1. Kernel command line
    5. Tools needed
        1. kernel build package
        2. dkms support
    6. Modify PKGBUILD
        1. prepare()
        2. _package-headers()
    7. Files Required
        1. certs-local/genkeys.py
        2. certs-local/x509.oot.genkey
        3. certs-local/install-certs.py
        4. certs-local/sign_manual.py
        5. certs-local/utils.py
        6. certs-local/signer_class.py
        7. certs-local/dkms/kernel-sign.conf
        8. certs-local/dkms/kernel-sign.sh
    8.  Arch AUR package

### See:

 - [Arch Wiki on signed kernel modules](https://wiki.archlinux.org/title/Signed_kernel_modules)
 - [Kernel Docs on module signing](https://www.kernel.org/doc/html/v6.2-rc2/admin-guide/module-signing.html)

# 1. Introduction 

The Linux kernel distinguishes and keeps separate the verification of modules from requiring or 
forcing modules to verify before allowing them to be loaded. Kernel modules fall into 2 classes:

Standard "in tree" modules which come with the kernel source code. They are compiled during the 
normal kernel build.

Out of tree modules which are not part of the kernel source distribution. They are built outside 
of the kernel tree, requiring the kernel headers package for each kernel they are to be built for. 
They can be built manually for a specific kernel and packaged, or they can be built whenever 
needed using [DKMS](https://wiki.archlinux.org/index.php/DKMS). 

Examples of such packages, provided by Arch, include:

  - [virtualbox-guest-modules-arch](https://www.archlinux.org/packages/?name=virtualbox-guest-modules-arch) 

During a standard kernel compilation, the kernel build tools create a private/public key pair and 
sign every in tree module (using the private key). The public key is saved in the kernel itself. 
When a module is subsequently loaded, the public key can then be used to verify that the module 
is unchanged.

The kernel can be enabled to always verify modules and report any failures to standard logs. 
The choice to permit the loading and use of a module which could not be verified can be either 
compiled into kernel or turned on at run time using a kernel parameter as explained below.
[Arch Wiki Kernel Parameters](https://wiki.archlinux.org/index.php/Kernel_parameter)

# 2. How to sign kernel modules using a custom kernel  

The starting point is based on a custom kernel package as outlined in this article 
[Kernel/Arch Build System](https://wiki.archlinux.org/index.php/Kernel/Arch_Build_System)

We will modify the build to:

  - Sign the standard in tree kernel modules
  - Provide what is needed to have signed out of tree modules and for the kernel to verify those modules.

Note: The goal here is to have:

  - In tree modules signed during the standard kernel build process.
    The standard kernel build creates a fresh public/private key pair on each build.

  - Out of tree modules are signed and the associated public key is compiled in to the kernel.
    We will create a separate public/private key pair on each build.

#     3. Summary of what needs to be done 

Each kernel build needs to made aware of the key/cert being used. Fresh keys are 
generated with each new kernel build.

A kernel config parameter is now used to make kernel aware of additional signing key: 

  CONFIG_SYSTEM_TRUSTED_KEYS="/path/to/oot-signing_keys.pem".

Keys and signing tools will be stored in current module build directory. Nothing needs to be done to 
clean this as removal is handled by the standard module cleanup. Certs are thus 
installed in /usr/lib/modules/<kernel-vers>-<build>/certs-local.  

#     4. Kernel configuration  

CONFIG_SYSTEM_TRUSTED_KEYS will be added automatically as explained below. 
In addition the following config options should be set by either manually editing the 
'config' file, or via make menuconfig in the linux 'src' area and subsequently copying 
the updated '.config' file back to the build file 'config'.  It is preferable to use elliptic curve type keys and zstd compression. 

  - CONFIG_MODULE_SIG=y
    Enable Loadable module suppot --->
    Module Signature Verification           -  activate

  - CONFIG_MODULE_SIG_FORCE=n
    Require modules to be validly signed -> leave off

        This allows the decision to enforce verified modules only as boot command line.
        If you are comfortable all is working then by all means change this to 'y'
        Command line version of this is : module.sig_enforce=1

  - CONFIG_MODULE_SIG_HASH=sha512
    Automatically sign all modules  - activate
    Which hash algorithm    -> SHA-512

  - CONFIG_MODULE_COMPRESS_ZSTD=y
    Compress modules on installation        - activate
        Compression algorithm (ZSTD)

  - CONFIG_MODULE_SIG_KEY_TYPE_ECDSA=y
    Cryptographic API --->
        Certificates for Signature Checking --->
            Type of module signing key to be generated -> ECDSA

  - CONFIG_MODULE_ALLOW_MISSING_NAMESPACE_IMPORTS=n
    Enable Loadable module support --->
        Allow loading of modules with missing namespace imports -> set off 

## 4.1 Kernel command line 

  After you are comfortable things are working well you can enable the kernel parameter to 
  require that the kernel only permit verified modules to be loaded:

    module.sig_enforce=1

# 5. Tools needed 

## 5.1 kernel build package 

In the directory where the kernel package is built:

      $ mkdir certs-local

This directory will provide the tools to create the keys, as well as signing kernel modules.

  - Copy these files into certs-local directory:

        genkeys.py
        x509.oot.genkey
        install-certs.py
        sign_module.py

### genkey.py & x509.oot.genkey

genkey.py along with its configuration file x509.oot.genkey are used to create key pairs.
It also provides the kernel with the key to sign out of tree modules by updating the config file 
used to build the kernel.

genkeys.py will create the key pairs in a directory named by date-time. It defaults to refreshing
the keys every 7 days but this can be changed with the *--refresh* command line option.

It also creates a soft link named 'current' which points to the newly created directory with the 'current' keys.
The actual key directory is named by date and time.

genkeys will check and update kernel configs given by the  --config config(s) option. This takes either a single
config file, or a shell glob for mulitple files. e.g. --config 'conf/config.*'. Remember to quote any wildcard 
characters to prevent the shell from expanding them. 
 
All configs will be updated with the same key. The default keytype is ec (elliptic curve) and the default
hash is sha512. These can be changed with command line options. See genkeys.py -h for more details.

###  sign_module.py 

signs out of tree kernel modules. It can be run manually but is typically invoked 
by dkms/kernel-sign.sh. It handles modules compressed with zstad, xz and gzip and depends on 
python-zstandard package to help handle those compressed with zstd. 


###  install-certs.py i

is called from the package_headers() function of PKGBUILD to install the signing keys. 
Example is given below. 
  
These files are all provided.

## 5.2 dkms support

    $ mkdir certs-local/dkms

Add 2 files to the dkms dir:

        kernel-sign.conf
        kernel-sign.sh

These will be installed in /etc/dkms and provide a mechanism for dkms to automatically sign 
modules (using the local key above) - this is the reccommended way to sign kernel modules. 
As explained, below - once this is installed - all that is needed to have dkms automatically 
sign modules is to make a soft link:

        $ cd /etc/dkms
        ln -s kernel-sign.conf <module-name>.conf

For example:

        ln -s kernel-sign.conf vboxdrv.conf

The link creation can easily be added to an arch package to simplify further if desired.

# 6. Modify PKGBUILD 

We need to make changes to kernel build as follows:

## 6.1 prepare()

Add the following to the top of the prepare() function:

    prepare() {

        msg2 "Rebuilding local signing key..."
        # adjust cerdir as needed 
        certdir='../certs-local'
        $certdir/genkeys.py -v --config ../config --refresh 30d
      ... 
    }

The default key regeneration refresh period is 7 days, but this can be changed on the command line. So if you want to create new keys monthly, then add "--refresh 30days" as an argument to genekeys.py. You can refresh on every build by using "--refresh always". Refresh units can be seconds,minutes,hours,days or weeks. 

## 6.2 _package-headers() 

  Add the following to the bottom of the _package-headers() function:

    _package-headers() {

    ...

    #
    # Out of Tree Module signing
    # This is run in the kernel source / build directory
    #
    msg2 "Local Signing certs for out of tree modules..."
      
    certs_local_src="../../certs-local" 
    certs_local_dst="${builddir}/certs-local"

    ${certs_local_src}/install-certs.py $certs_local_dst

    # install dkms tools if needed
    dkms_src="$certs_local_src/dkms"
    dkms_dst="${pkgdir}/etc/dkms"
    mkdir -p $dkms_dst

    rsync -a $dkms_src/{kernel-sign.conf,kernel-sign.sh} $dkms_dst/
    }

# 7. Files Required 

This is the list of files referenced above. Do not forget to make the scripts executable.

  - certs-local/genkeys.py
  - certs-local/install-certs.py
  - certs-local/x509.oot.genkey
  - certs-local/sign_module.py
  - certs-local/utils.py
  - certs-local/signer_class.py
  - certs-local/dkms/kernel-sign.conf
  - certs-local/dkms/kernel-sign.sh

# 8. Arch AUR packags

 - There is also an Arch AUR package [Arch Sign Modules](https://aur.archlinux.org/packages/arch-sign-modules) and its companion github repo [Arch-SKM](https://github.com/itoffshore/Arch-SKM) that leverages Arch-SKM. 

arch-sign-modules reduces the manual steps for building a fully signed custom kernel to 3 commands to *Update* / *Build* & *Install* the kernel:

    abk -u kernel-name
    abk -b kernel-name
    abk -i kernel-name

 - [README.scripts.md](https://github.com/itoffshore/Arch-SKM/blob/master/README.scripts.md)
 - [PKGBUILD example](https://github.com/itoffshore/Arch-SKM/blob/master/Arch-Linux-PKGBUILD-example)

## License

Created by Gene C. and licensed under the terms of the MIT license.
 - SPDX-License-Identifier: MIT  
 - Copyright (c) 2020-2023, Gene C



CHANGELOG.md
# Changelog

## [2.5.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-03-31
 - update changelog  
 - small README wordsmithing  
 - update CHANGELOG.md  

## [2.5.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-03-30
 - Use Full path to openssl executable /usr/bin/openssl  
 - Rename Changelog to CHANGELOG following our current release tools. Update  

## [2.4.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2023-01-06
 - Add SPDX licensing lines  
   Update kernel doc reference link  
 - Update kernel.org doc link to more recent kernel  

## [2.3.9] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-25
 - suppress a few silly pylint warnings  

## [2.3.8] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-21
 - Remove uneeded exception we just added. It is a subclass of OSError and so not needed.  

## [2.3.7] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-21
 - utils open_file catches (OSError,FileNotFoundError) exceoptions now  

## [2.3.6] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-18
 - update changelog  
 - Fix open_file to only apply encoding to text files (thanks  pylint)  

## [2.3.5] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-18
 - - More tidy up - try keep pylint noise down so dont miss things it finds  
 - delete old stuff  
 - comment tweak  

## [2.3.4] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-18
 - Update Changelog  
 - More code tidying in genkeys  
 - little more tidyup - no functional changes  
 - Add missing date to Changelog  

## [2.3.3] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-09
 - update Changelog  
 - Use OSError exception which has replaced IOError  
 - Catch OSError when file open fails  

## [2.3.2] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-08
 - Ack and Tested by by @itoffshore  
 - update Changelog  
 - trivial tidy  

## [2.3.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-08
 - more code tidying  
 - Update Changelog  
 - fix typo for refresh check  
 - tidy and improve exception handling  
 - tidy  
 - more cleaning  
 - more tidy  
 - more tidy ups  
 - some code tidying  
 - another typo!  
 - typo  
 - fix file to name to avoid module conflict  

## [2.3.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-08
 - - Code re-org to be more robust and easier to read.  
   - Introduce KernelModSigner class and ModuleTool class to help organize  
   - Functionality is unchanged.  

## [2.2.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-04
 - Update Changelog and README to reflect sign_module.py replacing sign_manual.sh  
 - Changelog - add date for 2.2.0  

## [2.2.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-04
 - update changelog  
 - archive sign_manual.sh  
 - turn off dev to ready for production  
 - Improve module signing scripts:  
   - sign_module.py replaces sign_manual.sh  
   - dkms/kernel_sign.sh updated accordingly  
   - install-certs updated accordingly  
   - adds dependency : python-zstandard for handling zst compressed modules  
 - README - small markdown tweaks  

## [2.1.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-03
 - update changelog  
 - typo  

## [2.1.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-03
 - update Changelog  
 - The key type and hash are now saved in files along side the keys. This allows the signing script to read them, and this means it no longer has hardcoded hash.  the sign script falls back on sha512 in case of previous key directory without a saved hash  
 - remove extraneous |  

## [2.0.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - update changelog  
 - word smith README  
 - fix markdown on last addition  

## [1.3.5] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - Update README and Changelog  
 - Add few more words about some available tooks by @itoffshore  

## [1.3.4] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - Update Changelog  
 - White space patches from @itoffshore  

## [1.3.3] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - Update Changelog  
 - Typo in echo found by @itoffshore  
 - Changelog udpate  
 - Add reference to @itoffshore aur package and github repo  

## [1.3.2] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - Fix hexdump typo "--e" to "-e"  
 - Changelog update  
 - Mindor markdown tweaks  

## [1.3.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - typo fix  
 - Update Changelog  

## [1.3.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-02
 - Per @ittoffshore, add comment about quoting wildcard characters  
 - Fixes from @itoffshore  
   1. For manual signing  
   zstd modules use .zst instead of .zsrd  
   support for gzip  
   2. For dkms  
   Add gzip support  

## [1.2.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-01
 - Expand help with reminder wildcards must be quoted  

## [1.1.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-01
 - tweak the prepare() example  
 - small word smithing  

## [1.0.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-01
 - remove debugging  

## [1.0.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-05-01
 - Update readme and changelog  
 - genkeys now handles multiple configs using shell glob with --config  
 - support utilities  
 - Rename tools to utils  
 - Share coupld functions via tools.py  
 - Add install-certs.py for use by package_headers() to simplify PKGBUILD  
 - Update package_headers() to remove reference to file no longer being created. Part of issue #3  
 - Add a little markdown to Changlelog.md  
 - Update changes for 0.8.0 and 0.8.1  

## [0.8.1] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-04-30
 - Remove references to now unused scripts  

## [0.8.0] ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 2022-04-30
 - fix typo  
 - Tidy up README  
 - As per itoffshore check for key exists prior to getting mtime. Fixes bug in check_refresh()  


certs-local/genkeys.py
#!/usr/bin/python
# SPDX-License-Identifier: MIT
# Copyright (c) 2020-2023 Gene C
"""
 Create new pub/priv key pair for signing out of tree kernel modules.
 This program must reside in the certs-local dir but can be run from any directory.
 It will use it's own path to locate the cert-local dir

 Each key pair is stored by date-time

 Args:
  refresh  - time before new keys are created. e.g. --refresh 24h
             default is 7days. Units may be abbreviateed and one of secs, mins, hours,
             days or weeks
  khash    - sets hash (default is sha512)
  ktype    - rsa or ec (default is ec)
  config   - config file to update with signing key. May contain wildcard
             e.g. --config config
                  --config ../configs/config.*

 (This tool replaces both of the older bash scripts:  genkeys.sh and fix_config.sh)
 NB:
   We always check the config - even if not refreshing keys to be sure it has the
   current signing key.

  Default refresh key is 7 days
"""
# Gene 2022-04-30
#
import os
import sys
import stat
import datetime
import argparse
import uuid
import re

#import pdb
import utils

#------------------------------------------------------------------------
class GenKeys :
    """
    Class to create out of tree kernel signing keys
    """
    # pylint: disable=R0902
    def __init__(self) :
        """
        Command line args and initialize
        """
        myname = os.path.basename(sys.argv[0])
        cert_dir = os.path.dirname(sys.argv[0])
        self.cert_dir = os.path.abspath(cert_dir)
        self.cwd = os.getcwd()

        if self.cwd == self.cert_dir:
            self.config = '../config'
        else:
            self.config = './config'

        self.verb = False
        self.refresh = '7d'
        self.khash = 'sha512'
        self.ktype = 'ec'
        self.kconfig_list = None
        self.okay = True

        par = argparse.ArgumentParser(description=myname)

        par.add_argument('-r',  '--refresh',
                default = self.refresh,
                help=f'key refresh period ({self.refresh}) E.g. "7d", "24h", "always"')

        par.add_argument('-c',  '--config',
                default = self.config,
                help=f'Kernel Config(s) to be updated ({self.config})'
                    + ' - wildcards ok (quote to avoid shell expansion).')

        par.add_argument('-kh', '--khash',
                default = self.khash,
                help=f'Hash type ({self.khash})')

        par.add_argument('-kt', '--ktype',
                default = self.ktype,
                help=f'Crypto algo ({self.ktype}) - Must be either ec or rsa')

        par.add_argument('-v',  '--verb',
                action='store_true', default = self.verb,
                help='Verbose ({self.verb})')

        parsed = par.parse_args()
        if parsed:
            for (opt,val) in vars(parsed).items() :
                setattr(self, opt, val)

        self.kconfig_list = utils.file_list_glob(self.config)
        if not self.kconfig_list :
            print ('No matching kernel config files found')
            self.okay = False

    def check_kern_config(self):
        """
        # Check kernel config(s) :
        # The config(s) checked to ensure hash and key_type match whats expected
        """
        # pylint: disable=R0914
        okay = True

        setting = {
                'rsa' : 'CONFIG_MODULE_SIG_KEY_TYPE_RSA',
                'ec'  : 'CONFIG_MODULE_SIG_KEY_TYPE_ECDSA',
                }
        to_match = {
                'ktype' : setting[self.ktype],
                'khash' : 'CONFIG_MODULE_SIG_HASH',
                }
        want_get = {
                'ktype' : 'y',
                'khash' : self.khash,
                }


        # check each config warn if mismatch
        num_to_match = len(to_match)
        num_with_errors = 0
        for kconfig in self.kconfig_list:
            fobj = utils.open_file(kconfig, 'r')
            if fobj:
                conf_items = fobj.readlines()
                fobj.close()
            else:
                print (f'Failed to open : {kconfig}')
                num_with_errors += 1
                continue

            count = 0
            for item in conf_items:
                for ckey,cval in to_match.items():
                    if item.startswith(cval):
                        count += + 1
                        want = want_get[ckey]
                        found = item.split('=')[1]
                        found = found.strip('"\n')
                        found = found.lower()
                        if found != want:
                            print (f'Bad {ckey} in {kconfig}  Want {want}, got {found}')
                            num_with_errors += 1

                    if count >= num_to_match:
                        # stop when get required config items
                        break

        if num_with_errors > 0:
            num_kconfigs = len(self.kconfig_list)
            print (f' {num_with_errors} out of {num_kconfigs} bad kernel config files')
            return not okay
        return okay

    def update_configs(self):
        """
        Update configs with new keys if needed
        Safest is to always read the current link and check config regardless if key was refreshed.
        """
        all_ok = True

        # ensure we get path to actuak directory and not the link name which doesn't change
        keydir = None
        keyname = 'signing_key.pem'
        keycur = os.path.join(self.cert_dir, 'current')

        if os.path.islink(keycur) :
            keydir = os.readlink(keycur)
            keydir = os.path.join(self.cert_dir, keydir)
            keydir = os.path.abspath(keydir)
            signing_key = os.path.join(keydir, keyname)
            #signing_key = os.path.normpath(signing_key)
            if not os.path.exists(signing_key):
                print(f'Failed to find signing key: {signing_key}')
                return not all_ok
            #
            # format to match RHS of kernel config file
            #
            signing_key = '"' + signing_key + '"\n'
        else:
            print (f'Missing : {keycur}')
            return not all_ok

        for kconfig in self.kconfig_list:
            kconfig_path = os.path.abspath(kconfig)
            okay = self._update_one_config(kconfig_path, signing_key)
            all_ok = all_ok | okay

        return all_ok

    def _update_one_config(self, kconfig_path, signing_key):
        """
        update a single kernel config
        """
        # pylint: disable=R0914
        okay = True
        kconfig_dir = os.path.dirname(kconfig_path)

        # always make temp in same dir to avoid rename across file systems
        kconfig_name_temp = str(uuid.uuid4())
        kconfig_path_temp = os.path.join(kconfig_dir, kconfig_name_temp)

        conf_name = 'CONFIG_SYSTEM_TRUSTED_KEYS='

        # read existing config
        fobj = utils.open_file(kconfig_path, 'r')
        if fobj:
            kconfig_rows = fobj.readlines()
            fobj.close()
        else:
            print (f'Failed to open : {kconfig_path}')

        changed = True
        new_rows = []
        for row in kconfig_rows:
            if row.startswith(conf_name):
                rsplit = row.split('=')
                cur_signing_key = rsplit[1]
                if cur_signing_key == signing_key:
                    changed = False
                    break
                new_row = conf_name + signing_key
                new_rows.append(new_row)
            else:
                new_rows.append(row)

        if changed:
            new_config = ''.join(new_rows)
            if self.verb:
                print (f'Updating config: {kconfig_path}')

            fobj = utils.open_file(kconfig_path_temp, 'w')
            if fobj:
                #for row in new_rows:
                #    fobj.write(row)
                fobj.write(new_config)
                fobj.close()
                os.rename(kconfig_path_temp, kconfig_path)
            else:
                print (f'Failed to write : {kconfig_path_temp}')
        else :
            if self.verb:
                print ('config up to date')

        return okay

    def _create_new_keys(self, kvalid, kx509, kprv, kkey, kcrt):
        """
        Make the actual keys - rsa or ec using openssl
        """
        # pylint: disable=R0913
        okay = True

        # new key
        cmd = f'/usr/bin/openssl req -new -nodes -utf8 -{self.khash} -days {kvalid}'
        cmd = cmd + f' -batch -x509 -config {kx509}'
        cmd = cmd + f' -outform PEM -out {kkey} -keyout {kkey}'

        if self.ktype == 'ec':
            cmd = cmd + ' -newkey ec -pkeyopt ec_paramgen_curve:secp384r1'

        pargs = cmd.split()
        [retc, _stdout, stderr] = utils.run_prog(pargs)
        if retc != 0:
            print('Error making new key')
            if self.verb and stderr:
                print(stderr)
            return not okay

        os.chmod (kkey, stat.S_IREAD|stat.S_IWRITE)

        # extract private key
        cmd = f'/usr/bin/openssl pkey -in {kkey} -out {kprv}'
        pargs = cmd.split()
        [retc, _stdout, stderr] = utils.run_prog(pargs)
        if retc != 0:
            print('Error making prv key')
            if self.verb and stderr:
                print(stderr)
            return not okay

        # Extract certificate (public) part
        cmd = f'/usr/bin/openssl x509 -outform der -in {kkey} -out {kcrt}'
        pargs = cmd.split()
        [retc, _stdout, stderr] = utils.run_prog(pargs)
        if retc != 0:
            print('Error making crt')
            if self.verb and stderr:
                print(stderr)
            return not okay
        return okay

    def make_new_keys (self):
        """
        Set up before we use openssl to create_new_keys()
        """
        # pylint: disable=R0914
        okay = True
        if self.verb:
            print ('Making new keys ')

        now = utils.date_time_now()
        now_str = now.strftime('%Y%m%d-%H%M')

        kdir = os.path.join(self.cert_dir, now_str)
        os.makedirs(kdir, exist_ok=True)

        kvalid = '36500'
        kx509 =  os.path.join(self.cert_dir, 'x509.oot.genkey')
        kbasename = 'signing'
        kprv = os.path.join(kdir, kbasename + '_prv.pem')
        kkey = os.path.join(kdir, kbasename + '_key.pem')
        kcrt = os.path.join(kdir, kbasename + '_crt.crt')

        khash_file = os.path.join(kdir, 'khash')
        ktype_file = os.path.join(kdir, 'ktype')

        okay = self._create_new_keys(kvalid, kx509, kprv, kkey, kcrt)
        if not okay:
            return not okay

        #
        # Put khash and ktype files in key dir.
        # The module signing script will read the hash/ktype
        #
        fobj = utils.open_file(khash_file,'w')
        if fobj:
            fobj.write(self.khash + '\n')
            fobj.close()
        else:
            print (f'Failed to write : {khash_file}')

        fobj = utils.open_file(ktype_file,'w')
        if fobj:
            fobj.write(self.ktype + '\n')
            fobj.close()
        else:
            print (f'Failed to open : {ktype_file}')

        # update current link to new kdir
        # since the 'current' link and the actual keydir are in same dir we use
        # relative for link - safest in case certs-local is moved
        if okay:
            link_temp = str(uuid.uuid4())
            link_temp = os.path.join(self.cert_dir, link_temp)

            kdir_rel = os.path.basename(kdir)
            cur_link = os.path.join(self.cert_dir, 'current')

            os.symlink(kdir_rel, link_temp)
            os.rename(link_temp, cur_link)

        return okay

    def check_refresh(self):
        """
        # check if key refresh is needed
        """
        okay = True
        if not self.refresh:
            return okay
        if self.refresh.lower() == 'always':
            return okay

        # get the refresh time
        parse = re.findall(r'(\d+)(\w+)', self.refresh)[0]
        if parse and len(parse) > 1:
            freq = int(parse[0])
            units = parse[1]
        else:
            print ('Failed to parse refresh string')
            return okay

        kfile = os.path.join(self.cert_dir, 'current', 'signing_key.pem')
        if os.path.exists(kfile) :
            mod_time = os.path.getmtime(kfile)
            curr_dt = datetime.datetime.fromtimestamp(mod_time)

            match units[0]:
                case 's':
                    timedelta_opts = {'seconds' : freq}
                case 'm':
                    timedelta_opts = {'minutes' : freq}
                case 'h':
                    timedelta_opts = {'hours' : freq}
                case 'd':
                    timedelta_opts = {'days' : freq}
                case 'w':
                    timedelta_opts = {'weeks' : freq}

            next_dt = curr_dt + datetime.timedelta(**timedelta_opts)
            now = utils.date_time_now()
            if next_dt > now:
                okay = False
        return okay

def main():
    """
    # genkeys - makes out-of-tree kernel module signing keys
    """
    #pdb.set_trace()
    genkeys = GenKeys()
    if not genkeys.okay:
        print ('Problem initializing')
        return 0

    if not genkeys.check_kern_config() :
        return 0

    if genkeys.check_refresh() :
        genkeys.make_new_keys()
    elif genkeys.verb:
        print ('Key refresh not needed yet')

    # always update to be sure config has key even if no refresh
    genkeys.update_configs()

    return 0

# ----------------------------
if __name__ == '__main__':
    main()
# ----------------------------
